### ISSUE TITLE
AmqpException during streaming 

### ISSUE THREAD

**Issue by sakthig** [Azure/azure-event-hubs-spark#367]
I am consuming IoThub data using spark.readStream. I keep getting these below errors

```
2018-08-01 07:17:37 WARN  TaskSetManager:66 - Lost task 1.1 in stage 9.0 (TID 1425, 172.31.14.123, executor 1): com.microsoft.azure.eventhubs.EventHubException: com.microsoft.azure.eventhubs.impl.AmqpException: The service was unable to process the request; please retry the operation. For more information on exception types and proper exception handling, please refer to http://go.microsoft.com/fwlink/?LinkId=761101 Reference:892b4926-d06f-4cc5-8ba8-9eb2f77a5b20, TrackingId:4bbd1e9d0002017000100a4f5b615e91_G1_B3, SystemTracker:ihsuprodmares019dednamespace:eventhub:iothub-ehub-sensinerve-355077-6a40c0307a~32766|analytics, Timestamp:8/1/2018 7:17:37 AM, errorContext[NS: ihsuprodmares019dednamespace.servicebus.windows.net, PATH: iothub-ehub-sensinerve-355077-6a40c0307a/ConsumerGroups/analytics/Partitions/1, REFERENCE_ID: 1b5455_94d2_G1_1533107847267, PREFETCH_COUNT: 2000, LINK_CREDIT: 2000, PREFETCH_Q_LEN: 0]
	at com.microsoft.azure.eventhubs.impl.ExceptionUtil.toException(ExceptionUtil.java:39)
	at com.microsoft.azure.eventhubs.impl.MessageReceiver.onClose(MessageReceiver.java:620)
	at com.microsoft.azure.eventhubs.impl.BaseLinkHandler.processOnClose(BaseLinkHandler.java:50)
	at com.microsoft.azure.eventhubs.impl.BaseLinkHandler.handleRemoteLinkClosed(BaseLinkHandler.java:70)
	at com.microsoft.azure.eventhubs.impl.BaseLinkHandler.onLinkRemoteClose(BaseLinkHandler.java:36)
	at org.apache.qpid.proton.engine.BaseHandler.handle(BaseHandler.java:176)
	at org.apache.qpid.proton.engine.impl.EventImpl.dispatch(EventImpl.java:108)
	at org.apache.qpid.proton.reactor.impl.ReactorImpl.dispatch(ReactorImpl.java:324)
	at org.apache.qpid.proton.reactor.impl.ReactorImpl.process(ReactorImpl.java:291)
	at com.microsoft.azure.eventhubs.impl.MessagingFactory$RunReactor.run(MessagingFactory.java:483)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
Caused by: com.microsoft.azure.eventhubs.impl.AmqpException: The service was unable to process the request; please retry the operation. For more information on exception types and proper exception handling, please refer to http://go.microsoft.com/fwlink/?LinkId=761101 Reference:892b4926-d06f-4cc5-8ba8-9eb2f77a5b20, TrackingId:4bbd1e9d0002017000100a4f5b615e91_G1_B3, SystemTracker:ihsuprodmares019dednamespace:eventhub:iothub-ehub-sensinerve-355077-6a40c0307a~32766|analytics, Timestamp:8/1/2018 7:17:37 AM
	... 13 more
```
Is this sequence of code causing this problem.

```
val agg1= stream
      .groupByKey(m => Stage1(m.customer))
      .mapGroupsWithState(GroupStateTimeout.ProcessingTimeTimeout())(stage1MappingFunction)

    agg1.collect()

    agg1
      .writeStream
      .outputMode(OutputMode.Update())
      .format("console")
      .start()
      .awaitTermination()
```

**Comment by sabeegrewal** [comment]
Hey @sakthig - this is a bug in `2.3.2`. If you run from `master` you shouldn't see this issue!

**Comment by jpsimen** [comment]
@sabeegrewal you are referring to the java.lang.AssertionError ("requestSeqNo does not match"), correct? 
Or is the AmqpException also a bug that is fixed already? We had the AmqpException when reading from one of our IotHubs this week.

**Comment by sabeegrewal** [comment]
@sakthig @jpsimen oh shoot, I'm just talking about the requestSeqNo issue.

**Comment by Fokko** [comment]
Doesn't `.collect()` and `.awaitTermination()` start two streams? I had similar issues when starting two streams on a single source.

**Comment by sabeegrewal** [comment]
From #374: 

Unable to instanciate EventHub sender for EventHub.

Caused by: com.microsoft.azure.eventhubs.EventHubException: com.microsoft.azure.eventhubs.impl.AmqpException: The service was unable to process the request; please retry the operation. For more information on exception types and proper exception handling, please refer to http://go.microsoft.com/fwlink/?LinkId=761101 TrackingId:a16b0a4dea5a4f55a3ea412950829118_G26, SystemTracker:gateway5, Timestamp:6/30/2018 2:42:39 AM
at com.microsoft.azure.eventhubs.impl.ExceptionUtil.toException(ExceptionUtil.java:39)
at com.microsoft.azure.eventhubs.impl.MessageSender.onClose(MessageSender.java:367)
at com.microsoft.azure.eventhubs.impl.BaseLinkHandler.processOnClose(BaseLinkHandler.java:50)
at com.microsoft.azure.eventhubs.impl.BaseLinkHandler.handleRemoteLinkClosed(BaseLinkHandler.java:70)
at com.microsoft.azure.eventhubs.impl.BaseLinkHandler.onLinkRemoteClose(BaseLinkHandler.java:36)
at org.apache.qpid.proton.engine.BaseHandler.handle(BaseHandler.java:176)
at org.apache.qpid.proton.engine.impl.EventImpl.dispatch(EventImpl.java:108)
at org.apache.qpid.proton.reactor.impl.ReactorImpl.dispatch(ReactorImpl.java:324)
at org.apache.qpid.proton.reactor.impl.ReactorImpl.process(ReactorImpl.java:291)
at com.microsoft.azure.eventhubs.impl.MessagingFactory$RunReactor.run(MessagingFactory.java:462)

**Comment by carrollwp** [comment]
[Driver_log_2.txt](https://github.com/Azure/azure-event-hubs-spark/files/2267556/Driver_log_2.txt)
[Driver_Log.txt](https://github.com/Azure/azure-event-hubs-spark/files/2267557/Driver_Log.txt)

Attaching databrick\spark logs. Able to reproduce with sample Databricks Docs code. 

18/08/07 16:16:09 INFO StateStore: Retrieved reference to StateStoreCoordinator: org.apache.spark.sql.execution.streaming.state.StateStoreCoordinatorRef@645fb740
18/08/07 16:16:09 INFO StateStore: Env is not null
18/08/07 16:16:09 INFO StateStore: Retrieved reference to StateStoreCoordinator: org.apache.spark.sql.execution.streaming.state.StateStoreCoordinatorRef@645fb740
18/08/07 16:16:09 INFO StateStore: Env is not null
18/08/07 16:16:09 INFO StateStore: Retrieved reference to StateStoreCoordinator: org.apache.spark.sql.execution.streaming.state.StateStoreCoordinatorRef@645fb740
18/08/07 16:16:39 INFO SessionHandler: entityName[zeiss-imt-cmmiot-iothub-d/ConsumerGroups/$Default/Partitions/2], sessionIncCapacity[0], sessionOutgoingWindow[2147483647]
18/08/07 16:16:39 INFO PartitionReceiverImpl: receiverPath[zeiss-imt-cmmiot-iothub-d/ConsumerGroups/$Default/Partitions/2], action[createReceiveLink], offset[null], sequenceNumber[160943], enqueuedTime[null], inclusiveFlag[true]
18/08/07 16:16:39 INFO ReceiveLinkHandler: linkName[7c63e6_d72e_G9_1533658599959], localSource[Source{address='zeiss-imt-cmmiot-iothub-d/ConsumerGroups/$Default/Partitions/2', durable=NONE, expiryPolicy=SESSION_END, timeout=0, dynamic=false, dynamicNodeProperties=null, distributionMode=null, filter={apache.org:selector-filter:string=UnknownDescribedType{descriptor=apache.org:selector-filter:string, described=amqp.annotation.x-opt-sequence-number >= '160943'}}, defaultOutcome=null, outcomes=null, capabilities=null}]
18/08/07 16:16:40 INFO ReceiveLinkHandler: linkName[7c63e6_d72e_G9_1533658599959], remoteSource[Source{address='zeiss-imt-cmmiot-iothub-d/ConsumerGroups/$Default/Partitions/2', durable=NONE, expiryPolicy=SESSION_END, timeout=0, dynamic=false, dynamicNodeProperties=null, distributionMode=null, filter={apache.org:selector-filter:string=org.apache.qpid.proton.codec.DecoderImpl$UnknownDescribedType@357cd1d}, defaultOutcome=null, outcomes=null, capabilities=null}]
18/08/07 16:16:40 INFO MessageReceiver: receiverPath[zeiss-imt-cmmiot-iothub-d/ConsumerGroups/$Default/Partitions/2], linkname[7c63e6_d72e_G9_1533658599959], updated-link-credit[2000], sentCredits[2000]
18/08/07 16:16:40 INFO BaseLinkHandler: linkName[7c63e6_d72e_G9_1533658599959], ErrorCondition[amqp:internal-error, The service was unable to process the request; please retry the operation. For more information on exception types and proper exception handling, please refer to http://go.microsoft.com/fwlink/?LinkId=761101 Reference:19dab54a-c4a7-4147-9fdd-f7299133084c, TrackingId:a55839b50000c5b3004774f85b69c5e8_G9_B14, SystemTracker:iothub-ns-zeiss-imt-301009-898b23a26d:eventhub:zeiss-imt-cmmiot-iothub-d~24575|$default, Timestamp:8/7/2018 4:16:40 PM]
18/08/07 16:16:40 INFO BaseLinkHandler: linkName[7c63e6_d72e_G9_1533658599959]
18/08/07 16:16:40 INFO SessionHandler: entityName[zeiss-imt-cmmiot-iothub-d/ConsumerGroups/$Default/Partitions/2], condition[Error{condition=null, description='null', info=null}]
18/08/07 16:16:40 ERROR Executor: Exception in task 2.1 in stage 76.0 (TID 695)
com.microsoft.azure.eventhubs.EventHubException: com.microsoft.azure.eventhubs.impl.AmqpException: The service was unable to process the request; please retry the operation. For more information on exception types and proper exception handling, please refer to http://go.microsoft.com/fwlink/?LinkId=761101 Reference:19dab54a-c4a7-4147-9fdd-f7299133084c, TrackingId:a55839b50000c5b3004774f85b69c5e8_G9_B14, SystemTracker:iothub-ns-zeiss-imt-301009-898b23a26d:eventhub:zeiss-imt-cmmiot-iothub-d~24575|$default, Timestamp:8/7/2018 4:16:40 PM, errorContext[NS: iothub-ns-zeiss-imt-301009-898b23a26d.servicebus.windows.net, PATH: zeiss-imt-cmmiot-iothub-d/ConsumerGroups/$Default/Partitions/2, REFERENCE_ID: 7c63e6_d72e_G9_1533658599959, PREFETCH_COUNT: 2000, LINK_CREDIT: 2000, PREFETCH_Q_LEN: 0]
                at com.microsoft.azure.eventhubs.impl.ExceptionUtil.toException(ExceptionUtil.java:39)
                at com.microsoft.azure.eventhubs.impl.MessageReceiver.onClose(MessageReceiver.java:620)
                at com.microsoft.azure.eventhubs.impl.BaseLinkHandler.processOnClose(BaseLinkHandler.java:50)
                at com.microsoft.azure.eventhubs.impl.BaseLinkHandler.handleRemoteLinkClosed(BaseLinkHandler.java:70)
                at com.microsoft.azure.eventhubs.impl.BaseLinkHandler.onLinkRemoteClose(BaseLinkHandler.java:36)
                at org.apache.qpid.proton.engine.BaseHandler.handle(BaseHandler.java:176)
                at org.apache.qpid.proton.engine.impl.EventImpl.dispatch(EventImpl.java:108)
                at org.apache.qpid.proton.reactor.impl.ReactorImpl.dispatch(ReactorImpl.java:324)
                at org.apache.qpid.proton.reactor.impl.ReactorImpl.process(ReactorImpl.java:291)
                at com.microsoft.azure.eventhubs.impl.MessagingFactory$RunReactor.run(MessagingFactory.java:483)
                at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
                at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
                at java.lang.Thread.run(Thread.java:748)
Caused by: com.microsoft.azure.eventhubs.impl.AmqpException: The service was unable to process the request; please retry the operation. For more information on exception types and proper exception handling, please refer to http://go.microsoft.com/fwlink/?LinkId=761101 Reference:19dab54a-c4a7-4147-9fdd-f7299133084c, TrackingId:a55839b50000c5b3004774f85b69c5e8_G9_B14, SystemTracker:iothub-ns-zeiss-imt-301009-898b23a26d:eventhub:zeiss-imt-cmmiot-iothub-d~24575|$default, Timestamp:8/7/2018 4:16:40 PM

**Comment by sabeegrewal** [comment]
Hey guys, this was fixed in #377 - please consume the new versions (2.2.3 or 2.3.3)! All stress runs complete run for days without running into this issue.

**Comment by xbaha** [comment]
I just got this error today, note I have not sent or received any messages when this happened, 
I am using Azure.Messaging.ServiceBus 7.4.0
it was running for 2 months with no problem, just happened suddenly.


```
Microsoft.Azure.Amqp.AmqpException: 
  | Description | "The service was unable to process the request; please retry the operation. For more information on exception types and proper exception handling, please refer to http://go.microsoft.com/fwlink/?LinkId=761101 TrackingId:1f66cacb98cb45ea99f0b85263c9d600_G1, SystemTracker:gateway7, Timestamp:2022-01-05T03:06:45" | string
```

```
  | StackTrace | "   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw() in /_/src/libraries/System.Private.CoreLib/src/System/Runtime/ExceptionServices/ExceptionDispatchInfo.cs:line 56\r\n   at Microsoft.Azure.Amqp.ExceptionDispatcher.Throw(Exception exception)\r\n   at Microsoft.Azure.Amqp.AsyncResult.End[TAsyncResult](IAsyncResult result)\r\n   at Microsoft.Azure.Amqp.AmqpObject.OpenAsyncResult.End(IAsyncResult result)\r\n   at Microsoft.Azure.Amqp.AmqpObject.EndOpen(IAsyncResult result)\r\n   at Microsoft.Azure.Amqp.AmqpObject.<>c.<OpenAsync>b__51_1(IAsyncResult r)\r\n   at System.Threading.Tasks.TaskFactory`1.FromAsyncCoreLogic(IAsyncResult iar, Func`2 endFunction, Action`1 endAction, Task`1 promise, Boolean requiresSynchronization) in /_/src/libraries/System.Private.CoreLib/src/System/Threading/Tasks/FutureFactory.cs:line 512\r\n--- End of stack trace from previous location ---\r\n   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw() in /_/src/libraries/System.Private.CoreLib/src/System/Runtime/ExceptionServices/ExceptionDispatchInfo.cs:line 56\r\n   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task) in /_/src/libraries/System.Private.CoreLib/src/System/Runtime/CompilerServices/TaskAwaiter.cs:line 173\r\n   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task) in /_/src/libraries/System.Private.CoreLib/src/System/Runtime/CompilerServices/TaskAwaiter.cs:line 150\r\n   at System.Runtime.CompilerServices.ConfiguredTaskAwaitable.ConfiguredTaskAwaiter.GetResult() in /_/src/libraries/System.Private.CoreLib/src/System/Runtime/CompilerServices/TaskAwaiter.cs:line 465\r\n   at Azure.Messaging.ServiceBus.Amqp.AmqpConnectionScope.<OpenAmqpObjectAsync>d__84.MoveNext() in /_/sdk/servicebus/Azure.Messaging.ServiceBus/src/Amqp/AmqpConnectionScope.cs:line 1122\r\n   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw() in /_/src/libraries/System.Private.CoreLib/src/System/Runtime/ExceptionServices/ExceptionDispatchInfo.cs:line 56\r\n   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task) in /_/src/libraries/System.Private.CoreLib/src/System/Runtime/CompilerServices/TaskAwaiter.cs:line 173\r\n   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task) in /_/src/libraries/System.Private.CoreLib/src/System/Runtime/CompilerServices/TaskAwaiter.cs:line 150\r\n   at System.Runtime.CompilerServices.ConfiguredTaskAwaitable.ConfiguredTaskAwaiter.GetResult() in /_/src/libraries/System.Private.CoreLib/src/System/Runtime/CompilerServices/TaskAwaiter.cs:line 465\r\n   at Azure.Messaging.ServiceBus.Amqp.AmqpConnectionScope.<OpenReceiverLinkAsync>d__69.MoveNext() in /_/sdk/servicebus/Azure.Messaging.ServiceBus/src/Amqp/AmqpConnectionScope.cs:line 365\r\n   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw() in /_/src/libraries/System.Private.CoreLib/src/System/Runtime/ExceptionServices/ExceptionDispatchInfo.cs:line 56\r\n   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task) in /_/src/libraries/System.Private.CoreLib/src/System/Runtime/CompilerServices/TaskAwaiter.cs:line 173\r\n   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task) in /_/src/libraries/System.Private.CoreLib/src/System/Runtime/CompilerServices/TaskAwaiter.cs:line 150\r\n   at System.Runtime.CompilerServices.ConfiguredTaskAwaitable`1.ConfiguredTaskAwaiter.GetResult() in /_/src/libraries/System.Private.CoreLib/src/System/Runtime/CompilerServices/TaskAwaiter.cs:line 551\r\n   at Azure.Messaging.ServiceBus.Amqp.AmqpReceiver.<OpenReceiverLinkAsync>d__35.MoveNext() in /_/sdk/servicebus/Azure.Messaging.ServiceBus/src/Amqp/AmqpReceiver.cs:line 200\r\n   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw() in /_/src/libraries/System.Private.CoreLib/src/System/Runtime/ExceptionServices/ExceptionDispatchInfo.cs:line 56\r\n   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task) in /_/src/libraries/System.Private.CoreLib/src/System/Runtime/CompilerServices/TaskAwaiter.cs:line 173\r\n   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task) in /_/src/libraries/System.Private.CoreLib/src/System/Runtime/CompilerServices/TaskAwaiter.cs:line 150\r\n   at System.Runtime.CompilerServices.ConfiguredTaskAwaitable`1.ConfiguredTaskAwaiter.GetResult() in /_/src/libraries/System.Private.CoreLib/src/System/Runtime/CompilerServices/TaskAwaiter.cs:line 551\r\n   at Microsoft.Azure.Amqp.FaultTolerantAmqpObject`1.<OnCreateAsync>d__6.MoveNext()" | string
```
